added bilinear() to tf in control lti

zolabar opened the pull request:
# bilinear discretization of continuous transfer function in Control API



#### References to other Issues or PRs

Fixes #24544


#### Brief description of what is fixed or changed

Added the function bilinear() to TransferFunction objects in sympy\physics\control\lti.py

It returns falling coeffs of H(z) from numerator and denominator. H(z) is the corresponding discretized tf, discretized with bilinear transform method. Coefficients are falling, i.e. H(z) = (az+b)/(cz+d) is returned as [a, b], [c, d].

#### Other comments

It is a useful function for numerical implementations of discrete signal processing and control with scipy.signal, control or matlab/simulink, once the symbolic continuous transfer function design is finished.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->

* functions
  * Added `bilinear`, a function for the bilinear transform discretization of a continuous transfer function from Control API.

<!-- END RELEASE NOTES -->

zolabar committed:
added bilinear() to tf in control lti

see

https://github.com/sympy/sympy/issues/24544

zolabar committed:
Update lti.py

zolabar committed:
Delete transfer_function_bilinear.ipynb

oscarbenjamin commented:
> The following commits **add new files**:
> 
> * [fe659ef](https://github.com/sympy/sympy/commit/fe659effc221103a8cfab9d13af59c5aa6bde071):
>   
>   * `examples/notebooks/transfer_function_bilinear.ipynb`
>   * `transfer_function_bilinear.ipynb`
> 
> The following commits **delete files**:
> 
> * [727fd6b](https://github.com/sympy/sympy/commit/727fd6b6a0166882a266e1aa34996da2932c2071):
>   
>   * `transfer_function_bilinear.ipynb`

These commits will need to be squashed. We can't have a commit that introduces a file followed by another that deletes it. The file must not exist in the history at all.

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -679,6 +683,52 @@ def __pow__(self, p):
+        T = Symbol('T')  # and sample period T
+       
+        H = self.num/self.den
+       
```
I see lot of blank lines in between the code which are not required.

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -679,6 +683,52 @@ def __pow__(self, p):
+        z = Symbol('z') # discrete variable z 
+        T = Symbol('T')  # and sample period T
+       
+        H = self.num/self.den
```
One workaround I see here which I would consider fundamentally incorrect is taking input in the form of a `TransferFunction` but then converting it into a symbolic expression and then working on it.

```
>>> from sympy import *
>>> from sympy.physics.control.lti import TransferFunction
>>> from sympy.abc import s, L, R, T, z
>>> tf = TransferFunction(1, s*L + R, s)
>>> type(tf)
<class 'sympy.physics.control.lti.TransferFunction'>
>>> H = tf.num/tf.den
>>> type(H)
<class 'sympy.core.power.Pow'>
```

This is not incorrect, but we are aware that `subs`, `collect`, `expand` do not work with the 'TransferFunction' class as of now. I would prefer that to be modified and then the entire function to be working on a variable of type - `<class 'sympy.physics.control.lti.TransferFunction'>`
Basically, if we are changing something then let's to it fully right. But if others agree then we can proceed with this too (it is clearly a useful addition).

zolabar replied:
@faze-geek , That proposal is realized in commit

[https://github.com/sympy/sympy/commit/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85](https://github.com/sympy/sympy/commit/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85)

thanks to comments, suggestions and contributions of @sylee957 and @hanspi42

faze-geek replied:
Thanks @zolabar , looks good to me. You can mark this conversation resolved accordingly.

faze-geek commented:
As far as the failing test is concerned you should run` python bin/strip_whitespace sympy/physics/control/lti.py`. Then your code quality check will pass.

zolabar committed:
python bin/strip_whitespace sympy/physics/control/lti.py

zolabar committed:
removed empty lines

zolabar committed:
flake8 test

zolabar committed:
corrected comments on example, z, T where unused

1e9abhi1e10 commented:
you need to add your name and email in the `.mailmap` file and then run `python bin/mailmap_check.py` for instructions read [here](https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file).

zolabar commented:
> you need to add your name and email in the `.mailmap` file and then run `python bin/mailmap_check.py` for instructions read [here](https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file).

thanks, just did that

zolabar commented:
Check the .mailmap file

> Am 21.01.2023 um 08:51 schrieb ABHISHEK PATIDAR ***@***.***>:
> 
> ﻿
> you need to add your name and email in the .mailmap file and then run python bin/mailmap_check.py for instructions read here.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

1e9abhi1e10 commented:
> Check the .mailmap file
> […](#)
> Am 21.01.2023 um 08:51 schrieb ABHISHEK PATIDAR ***@***.***>: ﻿ you need to add your name and email in the .mailmap file and then run python bin/mailmap_check.py for instructions read here. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.

After adding your name and email in the `.mailmap file` you need to run `python bin/mailmap_check.py` and then push a new commit

zolabar committed:
Update .mailmap

zolabar committed:
checks fail due to tar git synchronization, add this file later

zolabar commented:
> As far as the failing test is concerned you should run` python bin/strip_whitespace sympy/physics/control/lti.py`. Then your code quality check will pass.



> As far as the failing test is concerned you should run` python bin/strip_whitespace sympy/physics/control/lti.py`. Then your code quality check will pass.

Thanks, now all checks have passed :)

sylee957 commented on sympy/physics/control/lti.py:
```diff
@@ -302,6 +303,9 @@ def __new__(cls, num, den, var):
         else:
             raise TypeError("Unsupported type for numerator or denominator of TransferFunction.")
 
+
+
+
```
```suggestion
```

sylee957 replied:
```suggestion
```

zolabar committed:
Update sympy/physics/control/lti.py

Co-authored-by: (Lazard) S.Y. Lee  <sylee957@gmail.com>

zolabar committed:
Update .mailmap

zolabar committed:
Update .mailmap

faze-geek commented:
> Thanks, now all checks have passed :)

Could you add relevant examples in form of tests in https://github.com/sympy/sympy/blob/master/sympy/physics/control/tests/test_lti.py

zolabar commented:
> > Thanks, now all checks have passed :)
> 
> Could you add relevant examples in form of tests in https://github.com/sympy/sympy/blob/master/sympy/physics/control/tests/test_lti.py

Ok, i‘ll prepare some tests

zolabar committed:
added test for tf.bilinear()

zolabar commented:
> Could you add relevant examples in form of tests in https://github.com/sympy/sympy/blob/master/sympy/physics/control/tests/test_lti.py

@faze-geek,  done [e595caf](https://github.com/sympy/sympy/pull/24558/commits/e595caf477e53204489300ee219e7defcb6a3c6c)

faze-geek commented:
> @faze-geek, done [e595caf](https://github.com/sympy/sympy/pull/24558/commits/e595caf477e53204489300ee219e7defcb6a3c6c)

Thanks, I would appreciate if any maintainer of `control` module could review it. @akshanshbhatt Do you mind taking a look at this !?

hanspi42 commented:
I think two things should be changed:
1. The method should take `z` and `T` as input arguments.
2. This can be done without creating a symbolic expression internally, as @faze-geek says: if `N` is the maximum of the degrees of numerator and denominator, then multiplying numerator and denominator by `(z+1)**N` after substitution gives polynomials in `z`.  E.g., if the numerator polynnomial is `a*s**2+b*s+c`, and `N=3` because we have a third-order denominator, the new numerator polynomial is `a*(2/T)**2*(z-1)**2*(z+1) + b*(2/T)*(z-1)*(z+1)**2 + c*(z+1)**3`. This looks simple to code.

zolabar commented:
> I think two things should be changed:
> 
> 1. The method should take `z` and `T` as input arguments.
> 2. This can be done without creating a symbolic expression internally, as @faze-geek says: if `N` is the maximum of the degrees of numerator and denominator, then multiplying numerator and denominator by `(z+1)**N` after substitution gives polynomials in `z`.  E.g., if the numerator polynnomial is `a*s**2+b*s+c`, and `N=3` because we have a third-order denominator, the new numerator polynomial is `a*(2/T)**2*(z-1)**2*(z+1) + b*(2/T)*(z-1)*(z+1)**2 + c*(z+1)**3`. This looks simple to code.

@hanspi42, thank you for the feedback.

Do I understand correctly, that  concerning point 1, you propose that the function should be called as (for tf as TransferFunction, discrete_var the symbolic discrete variable (e.g. z) and sample_per the symbolic sample period) :

`
tf.blinear(discrete_var, sample_per)
`
?

Concerning point 2, I don't see how to take advantage of your analysis, that is, how should I get the list of the numerator and denominator coeffcients of the discretized transfer function?

Taking into account your point 1. it could look as follows. However, I as I wrote above, I have no idea of how to avoid the defining ` H = self.num/self.den` and then using SymPy to get the list of the numerator and denominator coeffcients of the discretized transfer function (which are needed for discrete implementation in practice then)...


`def tf.blinear(self, discrete_var, sample_per):


        z = discrete_var
        T = sample_per
      
        H = self.num/self.den
        HZ = H.subs(self.var, (2/T)*(z-1)/(z+1))

        num, den = fraction(HZ.simplify())

        HZnum = collect(expand(num, z), z)
        HZnum = Poly(HZnum, z)

        HZden = collect(expand(den, z), z)
        HZden = Poly(HZden, z)

        num_coefs = HZnum.coeffs()
        den_coefs = HZden.coeffs()

        return num_coefs, den_coefs

`

hanspi42 commented:
> Concerning point 2, I don't see how to take advantage of your analysis, that is, how should I get the list of the numerator and denominator coeffcients of the discretized transfer function?

For example: 
```
from sympy import *
s, z = symbols('s z')
T, a, b, c, d = symbols('T a b c d', real=True)

n = a*s**2+b*s+c
d = b*s**3+c*s**2+d*s+a
H = n/d
display(H)

np = n.as_poly(s).all_coeffs()
dp = d.as_poly(s).all_coeffs()

# This is the suggested code starting from having numerator and denominator
# coefficients np and dp
N = max(len(np), len(dp)) - 1
n1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(np[::-1], range(len(np))) ])
d1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(dp[::-1], range(len(dp))) ])

# This shows the transfer function
H1 = n1/d1
display(H1)

# This is the solution you are using
H2 = H.subs(s, 2/T*(z-1)/(z+1))
display(H2)

# This shows that it is the same; must be zero.
dH = (H1-H2).simplify()
display(dH)
```

hanspi42 commented:
> tf.blinear(discrete_var, sample_per)

Yes ;)

zolabar commented:
> > Concerning point 2, I don't see how to take advantage of your analysis, that is, how should I get the list of the numerator and denominator coeffcients of the discretized transfer function?
> 
> For example:
> 
> ```
> from sympy import *
> s, z = symbols('s z')
> T, a, b, c, d = symbols('T a b c d', real=True)
> 
> n = a*s**2+b*s+c
> d = b*s**3+c*s**2+d*s+a
> H = n/d
> display(H)
> 
> np = n.as_poly(s).all_coeffs()
> dp = d.as_poly(s).all_coeffs()
> 
> # This is the suggested code starting from having numerator and denominator
> # coefficients np and dp
> N = max(len(np), len(dp)) - 1
> n1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(np[::-1], range(len(np))) ])
> d1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(dp[::-1], range(len(dp))) ])
> 
> # This shows the transfer function
> H1 = n1/d1
> display(H1)
> 
> # This is the solution you are using
> H2 = H.subs(s, 2/T*(z-1)/(z+1))
> display(H2)
> 
> # This shows that it is the same; must be zero.
> dH = (H1-H2).simplify()
> display(dH)
> ```

@hanspi42 , thank you for the fast reply and the elegant compact code example, i‘ll test it and think about it!

zolabar committed:
compact version without simplify

only use of poly as suggested by @hanspi42 , @faze-geek, @sylee957,
mathematical formulae and code prototype by @hanspi42

zolabar commented:
> > tf.blinear(discrete_var, sample_per)
> 
> Yes ;)

@hanspi42 , commited the changes, thank you very much for the algebraic hints and the code prototype.
Since the discrete variable does not appear in the output coeff list I only included the sample period as input, hence the current function is used as

`
nz, dz = tf.bilinear(T)
`

sylee957 commented:
I think that general interface is good. 
But I'm not sure it should be added under methods because transfer function basic classes only supports arithmetic now

I think that we can pull out as function outside if we are not clear.

sylee957 commented:
I think that adding additional features are okay if it is outside, but only want to avoid adding inside objects as much as possible because of [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle)

zolabar commented:
> I think that general interface is good. But I'm not sure it should be added under methods because transfer function basic classes only supports arithmetic now
> 
> I think that we can pull out as function outside if we are not clear.

@sylee957 , thank you for the feedback.

For me it would be OK to inlcude `bilinear` as an outside function, like in

[https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html)

zolabar committed:
bilinear as function not as method from TransferFunction

zolabar commented:
@sylee957,

in commit

[https://github.com/sympy/sympy/pull/24558/commits/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85(https://github.com/sympy/sympy/pull/24558/commits/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85) 

`bilinear` is a function, used as

`bilinear(tf, T)`

similar to its numerical counterpart from sicipy.signal

hanspi42 commented:
It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.

zolabar committed:
included formula of bilinear transform in docstrings

zolabar commented:
> It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.

-> done

hanspi42 commented:
> > It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.
> 
> -> done

It looks OK, but does not appear in the documentation (preview here: https://output.circle-artifacts.com/output/job/1eb985ff-0719-4e5a-a222-ce27d1c7ee08/artifacts/0/doc/_build/html/index.html)

Unfortunately I do not know how to do this; does any of the other reviewers know?

sylee957 commented:
You need to add a directive `..autofunction` to `doc/src/modules/physics/control/lti.rst`

zolabar committed:
.. autofunction:: bilinear

zolabar commented:
> You need to add a directive `.. autofunction: bilinear` to `doc/src/modules/physics/control/lti.rst`

-> done

zolabar commented:
> You need to add a directive `.. autofunction: bilinear` to `doc/src/modules/physics/control/lti.rst`

-> done

zolabar commented:
> It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.

@hanspi42 , i included the docstring that @sylee957 suggested. How did you generate the test preview of the documentation? Thanks

hanspi42 commented:
> How did you generate the test preview of the documentation?

This is done by github. Look under "All checks have passed", there is an entry "Click here to see a preview of the documentation"

hanspi42 commented:
@sylee957: This now looks OK to me.

sylee957 commented:
I think that docs should be more readable because summary is too long and there are not many examples to figure out how it works for general cases.

sylee957 commented:
Also, I find that the release notes are misleading to `functions`, I'd correct them manually

oscarbenjamin commented on .mailmap:
```diff
@@ -1526,6 +1527,7 @@ wookie184 <wookie1840@gmail.com>
 wuyudi <wuyudi119@163.com>
 ylemkimon <ylemkimon@naver.com> <mail@ylem.kim>
 znxftw <vishnu2101@gmail.com>
+zolabar <raszoufine@gmail.com>
```
You have added two separate entries to the .mailmap file:
https://github.com/sympy/sympy/blob/4fb6b42dd0941f05178aa570edf9ef61162ce960/.mailmap#L138-L148
I'm going to presume that you want your name and email to be recorded as:
```
Zoufiné Lauer-Baré <raszoufine@gmail.com>
```

oscarbenjamin replied:
I'm changing this in #24775. Please comment there if you would like a different name/email.

zolabar replied:
Thank you
